core/src/main/scala/com/softwaremill/react/kafka_api_opt1/Consumer.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt1
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+import akka.NotUsed
+import akka.stream.Graph
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.Shape
+import akka.stream.SourceShape
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.GraphDSL
+import akka.stream.scaladsl.Source
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.consumer.OffsetAndMetadata
+import org.apache.kafka.common.TopicPartition
+
+object Consumer {
+
+  // FIXME those should probably be real case classes (or value classes)?
+  type CommitMsg = Map[TopicPartition, OffsetAndMetadata]
+  type CommitConfirmation = Future[CommitMsg]
+
+  case class ConsumerShape[K, V](
+      commit: Inlet[CommitMsg],
+      messages: Outlet[ConsumerRecord[K, V]],
+      confirmation: Outlet[CommitConfirmation]
+  ) extends Shape {
+    override def inlets: immutable.Seq[Inlet[_]] = immutable.Seq(commit)
+    override def outlets: immutable.Seq[Outlet[_]] = immutable.Seq(messages, confirmation)
+    override def deepCopy(): Shape = ConsumerShape(
+      commit.carbonCopy(),
+      messages.carbonCopy(),
+      confirmation.carbonCopy()
+    )
+    override def copyFromPorts(inlets: immutable.Seq[Inlet[_]], outlets: immutable.Seq[Outlet[_]]): Shape = {
+      require(inlets.size == 1)
+      require(outlets.size == 2)
+      ConsumerShape(
+        inlets.head.as[CommitMsg],
+        outlets(0).as[ConsumerRecord[K, V]],
+        outlets(1).as[CommitConfirmation]
+      )
+    }
+  }
+
+  trait Control {
+    def stop(): Unit
+  }
+
+  def record2commit: Flow[ConsumerRecord[_, _], Map[TopicPartition, OffsetAndMetadata], NotUsed] =
+    Flow[ConsumerRecord[_, _]]
+      .map { msg =>
+        Map(new TopicPartition(msg.topic(), msg.partition()) -> new OffsetAndMetadata(msg.offset()))
+      }
+
+  def batchCommit(maxCount: Int, maxDuration: FiniteDuration): Flow[CommitMsg, CommitMsg, NotUsed] =
+    Flow[Map[TopicPartition, OffsetAndMetadata]]
+      .groupedWithin(maxCount, maxDuration)
```

Comment:
I second @13h3r on a pue time-based commit API. Committing offset should not be at the same rate of receiving message, especially when commiting to Zookeeper with of course has not the same throughput that Kafka. Setting batch commit interval is actually a tradeoff between at-least-once semantic and pressure put on the system receiving the commit. So I don't see why the speed of downstream would dictate the pressure put on the system receiving the commit, these are separate things.

Also, with a pure time-based I dont think there can be a overflow issue as if we stick with the low-level Kafka API, you just keep the max offsets for each partition. This does not go well with the "commit per msg instead of offset" way of thinking of Option 3, but such a batchCommit could go well with low-level close-to-Kafka Option 1. And I think such things are important for high-performance Kafka use.